Source/QED/QuantumSyncEngineWrapper.H

```diff
@@ -22,33 +42,236 @@ using WarpXQuantumSynchrotronWrapper =
+
+        return has_event_happend;
+    }
+
+private:
+    //laser wavelenght is not used with SI units
+    const amrex::Real dummy_lambda{1.0};
+
+    const PicsarQuantumSynchrotronCtrl ctrl;
+    const picsar::multi_physics::lookup_1d<amrex::Real> lookup_table;
+};
+
+/**
+ * Functor to generate a photon via the Quantum Synchrotron process
+ * and to update momentum accordingly
+ */
+class QuantumSynchrotronGeneratePhotonAndUpdateMomentum
+{
+public:
+    /**
+     * Constructor acquires a reference to control parameters and
+     * lookup tables data.
+     * lookup_table uses non-owning vectors under the hood. So no new data
+     * allocations should be triggered on GPU
+     */
+    QuantumSynchrotronGeneratePhotonAndUpdateMomentum(
+        QuantumSynchrotronEngineInnards& r_innards):
+        ctrl{r_innards.ctrl},
+        lookup_table{
+            r_innards.cum_distrib_coords_1.size(),
+            r_innards.cum_distrib_coords_1.data(),
+            r_innards.cum_distrib_coords_2.size(),
+            r_innards.cum_distrib_coords_2.data(),
+            r_innards.cum_distrib_data.data()
+        }{};
+
+    /**
+     * Generates sampling (template parameter) photons according to Quantum Synchrotron process.
+     * It can be used on GPU.
+     * @param[in,out] px,py,pz momentum components of the lepton. They are modified (SI units)
+     * @param[in] ex,ey,ez electric field components (SI units)
+     * @param[in] bx,by,bz magnetic field components (SI units)
+     * @param[in] weight of the lepton (code units)
+     * @param[out] g_px,g_py,g_pz momenta of generated photons. Each array should have size=sampling  (SI units)
+     * @param[out] g_weight weight of the generated photons. Array should have size=sampling (code units)
+     */
+    template <size_t sampling>
+    AMREX_GPU_DEVICE
+    AMREX_FORCE_INLINE
+    void operator()(
+    amrex::Real* px, amrex::Real* py, amrex::Real* pz,
+    amrex::Real ex, amrex::Real ey, amrex::Real ez,
+    amrex::Real bx, amrex::Real by, amrex::Real bz,
+    amrex::Real weight,
+    amrex::Real* g_px, amrex::Real* g_py, amrex::Real* g_pz,
+    amrex::Real* g_weight) const
+    {
+        //[sampling] random numbers are needed
+        amrex::GpuArray<amrex::Real, sampling>
+            rand_zero_one_minus_epsi;
+        for(auto& el : rand_zero_one_minus_epsi) el = amrex::Random();
+
+        PicsarQuantumSynchrotronEngine::
+        internal_generate_photons_and_update_momentum(
+            *px, *py, *pz,
+            ex, ey, ez,
+            bx, by, bz,
+            weight, sampling,
+            g_px, g_py, g_pz,
+            g_weight,
+            dummy_lambda,
+            lookup_table,
+            ctrl,
+            rand_zero_one_minus_epsi.data());
+    }
+
+private:
+    //laser wavelenght is not used with SI units
+    const amrex::Real dummy_lambda{1.0};
+
+    const PicsarQuantumSynchrotronCtrl ctrl;
+    const picsar::multi_physics::lookup_2d<amrex::Real> lookup_table;
```

Comment:
Same comment as for `BreitWheeler`, should this be e.g. a pointer?

Reply:
see reply above

Reply:
Yes, I've changed this.